server/src/main/java/org/elasticsearch/transport/TransportMessageListener.java

```diff
@@ -30,6 +30,14 @@ default void onRequestReceived(long requestId, String action) {}
      */
     default void onResponseSent(long requestId, String action) {}
 
+    /**
+     * Called for every action response sent before the response has been passed to the underlying network implementation.
+     * @param requestId the request ID (unique per client)
+     * @param action the request action
+     * @param response response instance
+     */
+    default void onBeforeResponseSent(long requestId, String action, TransportResponse response) {}
```

Comment:
@DaveCTurner wdyt? this should be ok right? In prod the overhead is negligible I think.